Google api by lib

lzbrooks opened the pull request:
Major refactoring to use google api libraries instead of their "discovery-based API"

A lot of simplification of processes and proper functional splits in programming (such as `_dissect_message`)

Also switched to a more standardized subject (eg `"0,#fbc84a (1/2),Info", "0,#fbc84a (2/2),Testing"` 

Full suite of tests with 100% code coverage (to the best of my ability)

Production tested on eu.pythonanywhere.com and raspberry pi

More documentation needed, but functional for now

Commit:
Rough draft gmail api lib

Translate direct API queries to google api lib calls
Simplifies auth and url contruction

Commit:
Refactor google api class

Pull out all auth functionality to a superclass
Single purpose classes
Easier to read
Pulled out internet calls to wrapper functions for easier testing

Commit:
Update tests for new gmail api

Worked through all gmail api class related tests
Made sure they all work properly with the new code
Updated camphor tree api to also use new gmail api class
Tweaked gmail api class to match

Commit:
Cleaned up unused import in camphor tree api

Commit:
Add gmail auth tests and dummy data

Test all gmail auth functions
Add dummy data for credentials and token files

Commit:
Move crecs and token file paths to env vars

Still easily testable
Allows for preservation of interface functionality as is
Updated tests and README to match

Commit:
Add more some tests for gmail api class

Adjustments as necessary

Commit:
Add more tests for gmail api class

Small adjustments to said class logic to fix bugs found

Commit:
Update requirements

Upgrade versions
Add google api libraries

Commit:
Add tests for gmail message dissect parts function

Commit:
Add test coverage

Add test coverage config
Remove external library code from test coverage

Commit:
Add more tests for last bit of google api lib code

Add mock to match
Finished test coverage for google api lib code

Commit:
Refactor cloud loop api class for ease of testing

Add underscore to show internal functions
Pull out requests call to make it easier to mock

Commit:
Restore requests function and mock

Fix underscores in conftest mock fixtures

Commit:
Refactor rock block api class with internal functions

Ease of testing

Commit:
Refactor cloud loop api init

Simplify init
Remove helper function
Pull out get_payload from init
Fix tests

Commit:
Refactor cloud loop api for simplification

Pulled out funcitonaloty into helper functions

Commit:
Refactor hex decoding for ease of testing

Put back init statements
Simplify init
Add decode_hex_message calls where appropriate
Fix tests

Commit:
More refactoring of cloud loop api

Make funcitonality more classy

Commit:
Add tests for cloud loop api

Some tests still in progress

Commit:
Redo message chunking

- Have subject set to static size hex string
- Put original message subject as first messsage chunk text
- Fix tests
- Move int casting of max_message_size into config

Commit:
Add more cloud loop tests

- Fix up tests
- Refactor cloud loop helper function

Commit:
Add last tests to cloud loop api encoder

- Fix tests with new knowledge

Commit:
Add tests for cloud loop hex decode

Commit:
Fix cloud loop decoder for new subject format

- Add tests for new functionality
- Refactor decoder for simplicity
- Fix tests

Commit:
Add last tests for cloud loop api

- Covers all helper functions

Commit:
Add camphor tree api server option test

- test when server option is invalid (neither Satsuki nor Mei)

Commit:
Add tests for rock block api

- Refactor rock block api to make it more classy

Commit:
Add tests for config

- Minor todo cleanup

Commit:
Remove old google api classes and gmail auth flow

- Move gmail auth flow funcitonality to new gmail api class

Commit:
Refactor out gmail auth flow

- Fix up some tests

Commit:
Rename google_api_lib to google_api

- Fix mocks
- Fix tests

Commit:
Add tests for gmail auth flow

- Matched to gmail auth re_watch and refresh_with_browser tests
- Minor edits for bugs found

Commit:
Add gmail auth flow scripting

- Fix tests

Commit:
Make scripts executable

Commit:
Remove obsolete config options

- Google api keys etc now in google managed files

Commit:
Fix google call return

- Already JSON
- Add stub test for new error case

Commit:
Print more log for gmail pings

- Fix tests up a little

Commit:
Print more logs for gmail dissection

- Fix up other prints

Commit:
Remove duplicate gmail api calls

- Vestigial calls to gmail api that used to do something removed
- Fix tests

Commit:
Refactor out mailbox check to top level

- Fixes imports when running script

PeterDowdy commented on apis/cloud_loop_api.py:
```diff
@@ -18,161 +19,153 @@
+        return payload
+
+    def _email_to_contact_number(self, email_list):
+        email_list = [self._get_contact_number_for_email(email) if email in self.contacts.values()
```
You could probably simplify this lookup into `return [self.contacts.get(email, email) for email in self.contacts.values()`

lzbrooks replied:
Close, but no cigar.  .get() uses the dict key, not the dict value for pulling up dict items.  I'm trying to get a dict key by dict value here and it shows,,

lzbrooks replied:
Close :) (`_email_to_contact_number` is looking up keys by values)
Works on `_contact_number_to_email` tho so still a win!

PeterDowdy replied:
Ah, right you are there.

PeterDowdy commented on apis/google_api.py:
```diff
@@ -1,266 +1,207 @@
+        if 'historyId' in response_json:
+            print("GMail Re-Watch Success")
+        else:
+            print("Gmail Re-Watch Failure")
```
I recommend adding some kind of fallback here that will send an alert... maybe push an SMS using an SMS service like twilio?

PeterDowdy commented on apis/google_api.py:
```diff
@@ -1,266 +1,207 @@
-        self.message_to = Config.get_email(message_to)
+
+class GMailAPI(GMailAuth):
+    def __init__(self, message_to=None, message_from=None, message_subject=None, message_text=None):
```
I recommend only putting things in the constructor that are common to all the functions a class has - it's possible to create an invalid instance of the class otherwise (e.g. `GmailAPI().send_gmail_message()` won't do anything useful whereas if `message` is a parameter of `send_gmail_message`, then it's immediately clear that `GmailAPI().send_gmail_message()` is wrong)

PeterDowdy commented on apis/google_api.py:
```diff
@@ -1,266 +1,207 @@
-                print("Message From: " + message_from)
-                if size_in_bytes < int(self.max_message_size):
+                print(f"Max Message Size Allowed: {self.max_message_size}")
+                print(f"Current Message Size: {size_in_bytes}")
+                print(f"Message From: {message_from}")
+                if size_in_bytes < self.max_message_size:
+                    print("Valid: Message Under Max Size")
                     message_text = base64.urlsafe_b64decode(message_part['body']['data']).decode('utf-8')
-                if size_in_bytes > int(self.max_message_size) \
+                if size_in_bytes > self.max_message_size \
                         and message_from in Config.get_whitelist().values():
+                    print("Valid: Message In Whitelist And Over Max Size")
                     message_text = base64.urlsafe_b64decode(message_part['body']['data']).decode('utf-8')
```
There's a lot of useful diagnostic information here that's going to get rolled over by the limited stdout capacity of a heroku dyno; it would be a good failsafe to `logger.info` these to, say, google cloud storage so that you can review failures after the fact.

PeterDowdy commented on apis/rock_block_api.py:
```diff
@@ -14,66 +14,64 @@
+        if self.status_of_mailbox[2] == 1:
+            print("Message Received")
+        print("Size in bytes of message: " + str(self.status_of_mailbox[4]))
+        print("Number of Messages in Queue: " + str(self.status_of_mailbox[5]))
```
You should return this as a `@dataclass` or a dictionary to enable easier testing of this code.

PeterDowdy commented on config/config.py:
```diff
@@ -97,4 +91,4 @@ def get_google_sub():
     def get_max_message_size():
         if 'CAMPHOR_TREE_MAX_SIZE' in os.environ:
-            return os.environ['CAMPHOR_TREE_MAX_SIZE']
+            return int(os.environ['CAMPHOR_TREE_MAX_SIZE'])
```
maybe you could do a .get() with a default, e.g. `return int(os.environ.get("CAMPHOR_TREE_MAX_SIZE","200"))` to ensure this never returns `None`?

lzbrooks replied:
That is a great idea :D
I have gone through and implemented this advice throughout `config.py`

PeterDowdy commented on apis/camphor_tree_api.py:
```diff
@@ -12,36 +12,40 @@ def send_satellite_message(email, info_level, message_body, server_option):
         rock_block_message.send_cloud_loop_message()
-    if server_option == 'Mei':
+        send_status = 'Send Success'
+    elif server_option == 'Mei':
```
Have you considered putting multiple routes in your API to not need multiple modes in the one API call, e.g. call `/mei` or `/satsuki`, etc.?

lzbrooks replied:
Ah, yeah.  I originally kept everything to one endpoint to avoid using any session cookies when putting the website behind a password :sweat_smile: 

PythonAnywhere comes with password-locked access as a platform option instead, so I'll look into pulling apart the functionality on another branch o/

PeterDowdy reviewed (commented):
Looks good overall, I mainly have some comments about complexity. Have you considered using type annotations to help keep track of function returns? e.g. if you have a function like this:

```
def get_max_message_size():
  return os.environ.get("MAX_MESSAGE_SIZE")
```

and a function like this:

```
def check_max_message_size(message):
  if len(message) > get_max_message_size():
    raise ValueError(message, "greater than max size", get_max_message_size()
```

It's not clear if `get_max_message_size` or `check_max_message_size` should handle the `None`. If you do this:

```
from typing import Optional
def get_max_message_size() -> Optional[int]:
  return os.environ.get("MAX_MESSAGE_SIZE")
```

then a type-checker like mypy can provide a warning in `check_max_message_size` complaining that `get_max_message_size` might not be an int always.

Commit:
Simplify config env calls

- Instead of calling os.environ dict directly, use get() instead
- Move default values into optional get() parameter
- Fix tests

Commit:
Partial implementation of type hints

- for use with mypy
- minor refactoring

Commit:
Handle more None cases in cloud loop api

- Fix tests

Commit:
Add rest of api type hints

- Add notes for future refactoring
- Check tests

Commit:
Check logging in satsuki

- Add notes for later refactoring

Commit:
Fix logging verbosity

- Fix tests

Commit:
Add basic logging of message parse failure

- Better logging of this error to be added later
